tools/update_roles.php

```diff
@@ -0,0 +1,669 @@
+         FROM users_permission_categories_rel upc
+         LEFT JOIN permission_categories pc ON pc.id=upc.permission_category_id
+         WHERE upc.user_id=:userID AND pc.category=:role",
+        array(
+         'userID' => $userID,
+         'role'   => $role,
+        )
+    );
+
+    if (empty($roleID)) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * UserHasPermission
+ *
+ * Checks if a given user has a given permission
+ *
+ * @param int    $userID     the user ID
+ * @param string $permission the name of the permission
+ *
+ * @return boolean
+ */
+function userHasPermission($userID, $permission)
+{
+    $DB = Database::singleton();
+
+    $permission = $DB->pselect(
+        "SELECT upr.userID
+         FROM user_perm_rel upr
+         LEFT JOIN permissions p USING (permID)
+         WHERE upr.userID=:userID AND p.code=:permission",
+        array(
+         'userID'     => $userID,
+         'permission' => $permission,
+        )
+    );
+
+    if (empty($permission)) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * RoleHasPermission
+ *
+ * Checks if a given role contains a given permission
+ *
+ * @param string $role       the name of the role
+ * @param string $permission the name of the permission
+ *
+ * @return boolean
+ */
+function roleHasPermission($role, $permission)
+{
+    $rolePermissions = getRolePermissions($role);
+
+    if (in_array($permission, $rolePermissions)) {
+        return true;
+    }
+    return false;
+}
+
+/**
+ * RoleHasPermission
+ *
+ * Checks if a given role contains a given permission
+ *
+ * @return boolean
+ */
+function refresh()
+{
+    $DB = Database::singleton();
+
+    // Iterate over each user, update their roles based on their permissions
+    foreach (getUserIDs() as $user) {
+
+        $userID = $user['ID'];
+
+        echo "Updating roles for " . $user['Real_name'] . "\n";
+
+        // delete existing roles
+        $DB->delete(
+            'users_permission_categories_rel',
+            array('user_id' => $userID)
+        );
+
+        // recalculate their roles
+        $newRoles = getUserRolesByPermission($user['ID']);
+
+        // update their roles in the database
+        foreach ($newRoles as $role) {
+            echo "\tAdding " . $role . "\n";
+            $DB->insert(
+                'users_permission_categories_rel',
+                array(
+                 'user_id'                => $userID,
+                 'permission_category_id' => getRoleID($role),
+                )
+            );
+        }
+
+        echo "\n";
+    }
+}
+
+/**
+ * AddPermission
+ *
+ * Adds a permission to a role
+ * Updates users with that role
+ *
+ * @param string $role       the name of the role
+ * @param string $permission the name of the permission
+ *
+ * @return nothing
+ */
+function addPermission($role, $permission)
+{
+    $DB = Database::singleton();
+
+    echo "Adding the " . $permission . " permission to the " . $role .
```

Comment:
You're using `"`, not `'`. You don't need to terminate the string to include a variable. This can just be

`"Adding the $permission permission to the role $role" [...]`

Reply:
Done